golem-cli/src/service/component.rs

```diff
@@ -77,6 +90,145 @@ pub trait ComponentService {
+                .download_remote_file(component_file)
+                .await
+                .map(|f| vec![f]),
+            _ => Err(GolemError(format!("Unsupported scheme: {}", scheme))),
+        }
+    }
+
+    async fn load_local_file(
+        &self,
+        component_file: InitialComponentFile,
+    ) -> Result<Vec<LoadedFile>, GolemError> {
+        // if it's a directory, we need to recursively load all files and combine them with their target paths and permissions.
+        let source_path = PathBuf::from(component_file.source_path.as_url().path());
+
+        let mut results: Vec<LoadedFile> = vec![];
+        let mut queue: VecDeque<(PathBuf, ComponentFilePathAndPermissions)> =
+            vec![(source_path, component_file.target)].into();
+
+        while let Some((path, target)) = queue.pop_front() {
+            if path.is_dir() {
+                let read_dir = tokio::fs::read_dir(&path).await.map_err(|err| {
+                    GolemError(format!("Error reading directory {}: {err}", path.display()))
+                })?;
+                let mut read_dir_stream = ReadDirStream::new(read_dir);
+
+                while let Some(entry) = read_dir_stream.next().await {
+                    let entry = entry.map_err(|err| {
+                        GolemError(format!("Error reading directory entry: {}", err))
+                    })?;
+                    let next_path = entry.path();
+
+                    let file_name = entry.file_name().into_string().map_err(|_| {
+                        GolemError("Error converting file name to string".to_string())
+                    })?;
+
+                    let mut new_target = target.clone();
+                    new_target
+                        .extend_path(file_name.as_str())
+                        .map_err(|_| GolemError("Error extending path".to_string()))?;
+
+                    queue.push_back((next_path, new_target));
+                }
+            } else {
+                let content = tokio::fs::read(&path).await.map_err(|err| {
+                    GolemError(format!(
+                        "Error reading component file {}: {err}",
+                        path.display()
+                    ))
+                })?;
+                results.push(LoadedFile { content, target });
+            }
+        }
+        Ok(results)
+    }
+
+    async fn download_remote_file(
+        &self,
+        component_file: InitialComponentFile,
+    ) -> Result<LoadedFile, GolemError> {
+        let content = self
+            .file_download_client
+            .download_file(component_file.source_path.into_url())
+            .await
+            .expect("request failed");
+        Ok(LoadedFile {
+            content,
+            target: component_file.target,
+        })
+    }
+
+    async fn build_files_archive(
+        &self,
+        component_files: Vec<InitialComponentFile>,
+    ) -> Result<ComponentFilesArchiveAndProperties, GolemError> {
```

Comment:
I think I would rename it to `ComponentFilesArchiveWithProperties`. If its a `product` / `struct`, may be its just enough to say `ComponentFilesArchive` allowing it to hold more information in future, without the need to rename.